README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- badges: start -->

[![GitHub R package
version](https://img.shields.io/github/r-package/v/nicholascarey/loomeR)](https://github.com/nicholascarey/loomeR)
[![R-CMD-check](https://github.com/nicholascarey/loomeR/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/nicholascarey/loomeR/actions/workflows/R-CMD-check.yaml)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/nicholascarey/loomeR?branch=master&svg=true)](https://ci.appveyor.com/project/nicholascarey/loomeR)
[![Coverage
status](https://codecov.io/gh/nicholascarey/loomeR/branch/master/graph/badge.svg)](https://codecov.io/github/nicholascarey/loomeR?branch=master)
[![license](https://img.shields.io/badge/license-GPL--3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0.en.html)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1212570.svg)](https://doi.org/10.5281/zenodo.1212570)
<!-- badges: end -->

# Welcome

`loomeR` is an `R` package for creating looming animations for use in
behavioural and neurological experiments, and analysing escape
responses, for example the Apparent Looming Threshold (**ALT**) of an
escape (*Dill 1974, Webb 1982*). Looming animations are used in a range
of physiological, psychological and behavioural sciences to simulate an
approaching threat and investigate phenomena such as perception, visual
latency, escape responses, and neurological functioning. **ALT** is a
metric which describes the threshold where a specimen may initiate an
escape movement based on a combination of the perceived distance and/or
speed of an oncoming threat.

<p align="center">
<iframe src="https://player.vimeo.com/video/267812726" width="640" height="480" frameborder="0" webkitallowfullscreen mozallowfullscreen allowfullscreen>
</iframe>
</p>

See the
[Reference](https://nicholascarey.github.io/loomeR/reference/index.html)
page to view the detailed Help file for each function.

### Installation

`loomeR` is not yet published on CRAN, but can be installed using the
`devtools` package:

``` r
install.packages("devtools")
devtools::install_github("nicholascarey/loomeR")
```

### Usage

Using the package is straightforward:

#### 1. Create an animation model

``` r
diameter_model(), constant_speed_model(), variable_speed_model()
```

The package can create simple or complex animations in three ways:

- Use of basic inputs: enter start and end screen diameters, and total
  duration.
- Use of realistic parameters: the function will determine the correct
  screen diameters for each animation frame using…
  - **Constant speed**: specify a constant speed, size, and starting
    distance of the hypothetical oncoming threat
  - **Variable speed**: provide a profile of variable speeds, plus size
    and starting distance of the oncoming threat <br> <br>

#### 2. Create an animation from the model

``` r
looming_animation()
```

To create the animation from the model, `loomeR` requires
[ffmpeg](http://ffmpeg.org), a free, cross-platform, command-line
utility for encoding video to be installed on your system. This
currently works in `R` on **macOS** and **Windows**. Support for
**Linux** is planned: please [get in
touch](mailto:nicholascarey@gmail.com) if you would like to help with
testing on Linux systems.

Many options are available to customise the animation; specifying a
frame rate, modifying the colour and background, padding the video with
blank frames to a desired total duration, marking frames to assist with
identifying when escape responses occur, and more. <br> <br>

#### 3. Analyse escape responses

``` r
get_alt()
```

This function calculates the viewing angle, alpha (**α**), for each
frame in the animation, and the change in this viewing angle per unit
time (**dα/dt** in radians/second) (*Dill 1974*). Given a response frame
the *Apparent Looming Threshold* (**ALT**, *Webb 1982*) can be
determined. These metrics can all be corrected for different viewing
distances if the specimen has moved to a different distance from the
screen, which will affect the perceived **α** and thus **ALT**. The new
perceived speed and distance for the different viewing distance are also
returned. A visual response latency (i.e. to account for neurological
lag in response time) can also be applied. <br> <br>

### Example code

The included
[documentation](https://nicholascarey.github.io/loomeR/reference/index.html)
is comprehensive, and a vignette explaining how to use the package is in
preparation. For a quick evaluation try out the following code:

##### 1. Create a model

``` r
library(loomeR) # load the package
#> Loading required package: glue
#> Loading required package: plotrix
#> Loading required package: stringr

# Simple constant speed model
# (speed in cm/s, frame rate in Hz, all other inputs in cm)

x <- constant_speed_model(
  screen_distance = 20,    # How far from the screen is your observing specimen?
  frame_rate = 60,         # Frame rate you want the final animation to be 
  speed = 500,             # Speed of the simulated oncoming object
  attacker_diameter = 50,  # Diameter of the simulated oncoming object
  start_distance = 1000)   # Starting distance of the simulated oncoming object
```

##### 2. Use the model to create the animation

``` r
looming_animation(x)
```

##### 3. Extract the ALT

``` r
# E.g. from response frame 100, and applying a response latency of 60 milliseconds
get_alt(x, response_frame = 100, latency = 0.06)
#> 
#> Extraction complete. 
#>  
#> Response Frame:            100
#> Response Frame Adjusted:   96
#> Latency Applied:           0.06s
#>  
#> The Apparent Looming Threshold is: 
#> ALT: 0.5911 radians/sec
```

### Forthcoming and potential features

- [x] Function to extract model parameters at a particular frame,
  optionally apply a latency correction, and correct the perceived speed
  and distance for a different viewing distance.
  - **DONE** - see `get_alt`
- [x] Option to export images only, and not convert them
  - Partially done. See `save.images` input in `looming_animation()`.
- [ ] Linux support
- [ ] Add a receding option for animations (possibly works by setting
  speed as a negative, or start/end diameters the other way round, but
  this has not been tested)
- [ ] Enhancements (e.g. quicker padding method) and alternative to
  `ffmpeg`
- [ ] Option to subsample or interpolate variable speed profiles to
  match a desired frame rate
- [ ] Use of custom shapes (please [contact
  me](mailto:nicholascarey@gmail.com) if you know of actual applications
  for this… it would be a *lot* of work)

### Feedback

If you have any bugs or feedback, you can contact me via
[email](mailto:nicholascarey@gmail.com), or by [opening an
issue](https://github.com/nicholascarey/loomeR/issues).

### Acknowledgements

Working with the following people inspired the creation of this package:

- Paolo Domenici, CNR IAMC, Italy.
  [Link](http://iamc.objectis.net/staff/domenici-paolo)
- Jeremy Goldbogen & Dave Cade, Hopkins Marine Station, Stanford
  University. [Link](http://goldbogen.stanford.edu)
- Januar Harianto, University of Sydney
  [Link](https://github.com/januarharianto)

### Cite

If you use `loomeR` in your work, a citation using this zenodo DOI would
be much appreciated:

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1212570.svg)](https://doi.org/10.5281/zenodo.1212570)

Or run this code:

``` r
citation("loomeR")
#> 
#> To cite package 'loomeR' in publications use:
#> 
#>   Carey N (2022). _loomeR: Looming Animations for Use in Behavioural
#>   and Neurological Experiments_.
#>   https://github.com/nicholascarey/loomeR,
#>   https://doi.org/10.5281/zenodo.1212570.
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Manual{,
#>     title = {loomeR: Looming Animations for Use in Behavioural and Neurological Experiments},
#>     author = {Nicholas Carey},
#>     year = {2022},
#>     note = {https://github.com/nicholascarey/loomeR, https://doi.org/10.5281/zenodo.1212570},
#>   }
#> 
#> ATTENTION: This citation information has been auto-generated from the
#> package DESCRIPTION file and may need manual editing, see
#> 'help("citation")'.
```

Please do **[let me know](mailto:nicholascarey@gmail.com)** if you use
it in your experiments. I would love to keep an updated list of studies
which have made use of it, and I can help publicise your paper by
tweeting about it!

### References

Dill, Lawrence M, 1974. The escape response of the zebra danio
(Brachydanio rerio) I. The stimulus for escape. Animal Behaviour 22,
711–722. <https://doi.org/10.1016/S0003-3472(74)80022-9>

Webb, P.W., 1982. Avoidance responses of fathead minnow to strikes by
four teleost predators. J. Comp. Physiol. 147, 371–378.
<https://doi.org/10.1007/BF00609671>

Gibson, J. J. (2014) The Ecological Approach to Visual Perception:
Classic Edition. Psychology Press. (2014).

Domenici, P. (2002). The visually mediated escape response in fish:
predicting prey responsiveness and the locomotor behaviour of predators
and prey. Marine and Freshwater Behaviour and Physiology, 35(1–2),
87–110. <https://doi.org/10.1080/10236240290025635>

Muijres, F. T., Elzinga, M. J., Melis, J. M., & Dickinson, M. H. (2014).
Flies evade looming targets by executing rapid visually directed banked
turns. Science, 344(6180), 172–177.
<https://doi.org/10.1126/science.1248955>

Peron, S., & Gabbiani, F. (2009). Spike frequency adaptation mediates
looming stimulus selectivity in a collision-detecting neuron. Nature
Neuroscience, 12, 318. <http://dx.doi.org/10.1038/nn.2259>


NEWS.md
## loomeR 0.4.0 -- 2022-12-07

- `looming_animation` now checks that the output video has the correct number of frames, and if not gives a warning. This was in response to a user who had discovered their video had dropped frames. Likely this was because they were using an old PC, where - presumably - the hard disk write speeds were too slow to keep up with the images output by the function. See new `pause` input in next change for a solution if this happens. Note this warning is experimental - I have not been able to thoroughly test it on Windows. If you see it, it may not necessarily be true. 

- `looming_animation` has a new `pause` input. This is for users on slow PCs where frames may be dropped because the filesystem is overloaded by having to write so many image files (see above). This operator adds a delay in seconds to the generation of each frame. Generally it should be a small value, but enough that your system doesn't lag behind. For example, with a 600 frame video, 'pause = 0.05' will add an *additional* 30 seconds (600 * 0.05) to the total time taken.

- New `save_images` logical input in `looming_animation` and `looming_animation_calib` to save image files, not delete them. This is useful if you want to create the animation in an application other than `ffmpeg`, or run them through `ffmpeg` yourself.

## loomeR 0.3.0 -- 2019-07-09

- Models from `constant_speed_model` and `variable_speed_model` now contain alpha and da/dt columns calculated for the hypothesised screen viewing distance used to create them. (`diameter_model` objects, where these can't be calculated since there is never a viewing distance specified, also have these columns but they are filled with `NA`).

- Models from `constant_speed_model` and `variable_speed_model` are now slightly changed to end on last row with distance = 0. Just seems more correct to do that, however it means the first distance in row 1 isn't quite at the set `start_distance` (in `constant_speed_model` at least) but a single frame later at the specified speed. Hey, derivatives ¯\_(ツ)_/¯

- Added three internal functions to clean up code (`calc_alpha`, `calc_dadt`, `calc_screen_diam`)

#### `looming_animation` 

- Frame numbers can have custom text added to the start of each using `frame_number_tag`. E.g. A-1, A-2, or 1-1, 1-2, etc. This is super useful if you are creating multiple animations and want to make sure you don't get them mixed up.

- Output videos can have custom names using the `filename` operator. Note, the function will still
overwrite any videos it encounters with an identical filename.

- Distances from the screen edge of frame numbers and marker dots can be changed with the `_inset` operators for each

- Animations can now be looped with the `loop` operator. If this is set to 2 or greater, an additional video file is exported with the animation (including padding, if used) looped this number of times. This way you can create a long video containing multiple (though identical) looms separated by whatever time period you want (set by `pad`). For creating videos with multiple, *different* looms, I am working on an internal solution, but for now it's fairly easy to create multiple videos and join them with ffmpeg or another video editor.

- Padding frames now have their own frame numbers (if `frame_numbers = TRUE`), which are appended with a`P`. These are separate from animation frame numbers, and are specific to padded frames, i.e. they start at zero and end at the number of the final padding frame, followed by the actual animation with frame numbers that start at zero and use their own sequence. This is to assist in experiments where the animation is looped or repeated multiple times, and the user wants to identify the particular moment something happens outside of the animation playback period. For example, when a previously startled specimen returns to the experimental arena. They inherit the same `frame_number_tag` and other characteristics from the frame number settings.


## loomeR 0.2.2 -- 2019-02-05

A couple of important bug fixes related to padding videos. Also lots of minor code changes and also added unit testing code. 

v0.2.2 Changes
- FIX: making animation from a `diameter_model` would fail when padded.
- FIX: padding would sometimes fail because number of frames being added was not an integer.
- `citation("loomeR")` should return a citation. 
- Added unit testing. Most code now covered by `testthat` checks.
- Added numerous console messages and input checks to help with testing.
- Moved internal functions to separate script and removed unused ones.
- Minor changes to pass build checks and prepare for CRAN submission. 
- Minor changes to `get_alt` console print out.
- Revised README to separate code chunks, add `citation()` code, and a hit counter badge.
- Added some URLs to DESCRIPTION.
- Added macOS to Travis build checks.
- Added Appveyor Windows build checks.


## loomeR 0.2.1 -- 2019-01-11

Incorporated some internal testing using `testthat`. No outward facing changes.


## loomeR 0.2.0 -- 2018-05-02

This release introduces the `get_alt()` function. This lets you calculate the Apparent Looming Threshold for a particular response frame, as well as other metrics to analyse escape responses. Also includes various bugfixes and enhancements. 


## loomeR 0.1.3 -- 2018-04-13

### `diameter_model` updated:

- Fix for start diameter not quite correct under `constant_speed` expansion option
- Output no longer includes arbitrary `$distance` vector in `$model`
- Code cleaned up to remove use of arbitrary predator distance and size. 

- New internal fn for shortening vectors - `vect_rm_end`


## loomeR 0.1.2 -- 2018-04-05

Refinement release. Removed requirement for model object input in `looming_animation_calib`


## loomeR 0.1.1 -- 2018-04-05

Maintenance release. Minor edits to documentation.


## loomeR 0.1.0 -- 2018-04-05

Initial release.


tests/testthat.R
library(testthat)
library(loomeR)

test_check("loomeR")


tests/testthat/test-get_alt.R
capture.output({  ## stops printing outputs on assigning

  # test --------------------------------------------------------------------

  test_that("runs with diameter_model", {
    expect_output(str(get_alt(diameter_model(),
                              response_frame = 10,
                              new_distance = 20)),
                  "List of 13")
  })

  test_that("runs with constant speed model", {
    expect_output(str(get_alt(constant_speed_model(),
                              response_frame = 10)),
                  "List of 13")
  })

  test_that("runs with constant speed model", {
    expect_output(str(get_alt(variable_speed_model(x = c(1:100)),
                              response_frame = 10)),
                  "List of 13")
  })

  test_that("stops with wrong classes", {
    lst <- list(c(1:10), c(11:20))
    expect_error(get_alt(lst),
                 "Input must be an object of class 'constant_speed_model', 'variable_speed_model'")
  })

  test_that("stops with missing response_frame", {
    mod <- diameter_model(start_diameter = 10,
                          end_diameter = 9.5,
                          duration = 1,
                          frame_rate = 30)

    expect_error(get_alt(mod, response_frame = NULL),
                 "A 'response_frame' is required to extract the ALT and associated data.")
  })

  test_that("tops with response_frame = 1", {
    mod <- diameter_model(start_diameter = 10,
                          end_diameter = 9.5,
                          duration = 1,
                          frame_rate = 30)

    expect_error(get_alt(mod, response_frame = 1),
                 "ALT cannot be extracted from the first frame")
  })

  test_that("stops if response_frame out of range", {
    mod <- diameter_model(start_diameter = 10,
                          end_diameter = 9.5,
                          duration = 1,
                          frame_rate = 30)

    expect_error(get_alt(mod, response_frame = 31),
                 "The 'response_frame' is greater than the last frame of the animation model.")
  })

  test_that("stops if diameter_model and no new_distance entered", {
    mod <- diameter_model(start_diameter = 10,
                          end_diameter = 9.5,
                          duration = 1,
                          frame_rate = 30)

    expect_error(get_alt(mod,
                         response_frame = 20,
                         new_distance = NULL),
                 "Extracting data from 'diameter_model' objects requires a screen viewing distance")
  })

  test_that("class", {
    result <- get_alt(variable_speed_model(x = c(1:100)),
                      response_frame = 10)

    expect_is(result, "get_alt")
  })

  test_that("can print result", {
    result <- get_alt(variable_speed_model(x = c(1:100)),
                      response_frame = 10)

    expect_output(print(result))
  })

}) ## turns printing back on


tests/testthat/test-diameter_model.R
capture.output({  ## stops printing outputs on assigning

  test_that("list is created", {
    expect_output(str(diameter_model()), "List of 6")
  })

  test_that("class", {
    model <- diameter_model()
    expect_is(model, "diameter_model")
  })

  test_that("number of frames generated matches duration at frame rate", {
    # random duration and frame rate
    d <- round(runif(1, min=0, max=20), 2)
    f <- floor(runif(1, min = 12, max = 120))
    # create model
    mod <- diameter_model(duration = d,
                          frame_rate = f)
    # check number of frames output equals duration times frame rate rounded up
    expect_equal(nrow(mod$model),
                 ceiling(mod$duration*mod$frame_rate))
  })

  test_that("stops if expansion operator not set correctly", {
    expect_error(diameter_model(expansion = "diam_mod"),
                 "expansion operator not set correctly: must be 'constant_speed' or 'constant_diameter'")
  })

  test_that("works with expansion = 'constant_diameter'", {
    expect_output(str(diameter_model(expansion = "constant_diameter")),
                  "List of 6")
  })

}) ## turns printing back on


R/get_alt.R
#' @title Determine the Apparent Looming Threshold (ALT) for a given response
#'   frame
#'
#' @description `get_alt` returns the Apparent Looming Threshold (ALT), the rate
#'   of change of the visual angle (alpha, or **a**) of a looming shape
#'   (**da/dt** in radians/s), at a specified frame of the animation. Typically
#'   this is the 'response frame' - the frame at which a specimen responds to
#'   the stimulus.
#'
#'   The model originally used to create the looming animation must be supplied,
#'   along with a specified `response_frame`. The function returns the ALT at
#'   this frame. It also returns the perceived speed and perceived distance of
#'   the looming shape at this frame, as specified by the original model.
#'
#'   The function also allows the results to be 'corrected' in two ways:
#'
#'   - Different viewing distance: If the observing specimen has moved, or is
#'   otherwise at a different distance from the screen as that specified in the
#'   original model, this will affect the perceived ALT, as well as perceived
#'   distance and speed. To account for this, a `new_distance` can be specified,
#'   and the function will correct the ALT for the new viewing distance, as well
#'   as calculate the new perceived speed and distance.
#'
#'   - Visual response latency: The function also allows a visual response
#'   latency to be applied (`latency`, in s). Typically, there is a lag between
#'   an animal responding (voluntarily or involuntarily) to a stimulus, and the
#'   signal to move reaching their musculature. If this latency value is known
#'   or has been quantified, it can be entered as `latency` in seconds and the
#'   results will be returned from the closest matching frame to the
#'   `response_frame` minus this duration (see Details).
#'
#'
#' @details Note, that since **da/dt** is a derivative, its value lies *between*
#'   frames. The function returns the **da/dt** value (rate of change in the
#'   viewing angle) between the given `response_frame` and the previous frame.
#'   If, for whatever reason, you want the **da/dt** between the
#'   `response_frame` and the next frame, simply add 1 to the response frame.
#'
#'   Latency: Latency is applied internally by adjusting the response frame. The
#'   frame rate multiplied by the latency gives the number of frames 'backwards'
#'   from the entered `response_frame` the function goes to extract the results.
#'   In the event the number of frames backwards is not an integer, this is
#'   rounded. For example, with a 60 fps animation and 0.06s latency, the
#'   function would look to go back 3.6 frames; this is rounded to 4.
#'
#'   Diameter model: Note that for a [diameter_model()], the ALT (i.e. **da/dt**
#'   at a particular frame) can be extracted given a viewing distance, but *not*
#'   a perceived distance and speed. This is because the hypothetical size of
#'   the attacker was never specified in creating the model. While the viewing
#'   angle (**a**) and its derivative (**da/dt**) can be calculated, the
#'   perceived distance and/or speed cannot; a small object on screen expanding
#'   rapidly could represent a very small object moving slowly at a close
#'   distance, or a very large object moving rapidly at a far distance. Both of
#'   these can produce an identical **da/dt**, but without a size being
#'   specified the speed and distance cannot be determined. Of course,
#'   expressing the ALT as a **da/dt** value is designed to negate this exact
#'   issue, by focussing purely on the rate of change of the viewing angle. If
#'   these perceived speed and distance parameters may be important to your
#'   study however, the model should be created in [constant_speed_model()] or
#'   [variable_speed_model()]. Note, [diameter_model()] is intended to produce
#'   simple animations purely to induce a response where the precise parameters
#'   are not that important.
#'
#'
#'
#' @seealso [constant_speed_model()], [variable_speed_model()],
#'   [diameter_model()], [looming_animation()]
#'
#' @param x numeric. Object of class `constant_speed_model`,
#'   `variable_speed_model`, or `diameter_model`.
#' @param response_frame integer. The frame at which you want to determine the
#'   ALT.
#' @param new_distance numeric. Distance in cm the specimen is from the screen,
#'   if this is different from that used to create the model.
#' @param latency numeric. Visual response latency in seconds.
#'
#' @return The function returns a `list` object of class `get_alt`. The first
#'   value in the output object (`$alt`) is the ALT in rad/s at the response
#'   frame (with latency correction if entered). It also includes the original
#'   looming animation model (`$original_model`), and an adjusted model for any
#'   `new_distance`  (`$adjusted_model`). It also returns perceived speed and
#'   distance (`$speed_perceived`, `$distance_perceived`) at the response frame
#'   (with latency correction) adjusted for a new viewing distance, and also the
#'   original speed and distance in the model at that same frame without this
#'   correction (`$speed_in_model`, `$distance_in_model`). All angular and
#'   **da/dt** results are returned in radians, as is typically used in the
#'   literature. However the output also includes these in degrees with a `_deg`
#'   suffix (e.g. `$alt` vs. `$alt_deg`). All calculated (**a** & **da/dt**) and
#'   adjusted (perceived speed and distance) parameters are available for every
#'   frame in the `$adjusted_model$model` component. Note, that because
#'   **da/dt** is a derivative, the **da/dt** vector returned in
#'   `adjusted_model$model` will be one value shorter than the total number of
#'   frames, and therefore starts with a blank value (i.e. `NA`). During
#'   processing, values of **da/dt** are determined from the previous frame to
#'   the current frame. Therefore a **da/dt** value for a particular frame is
#'   how the angle changed from the previous frame to that frame.
#'
#' @examples
#' ## create looming animation model
#' loom_model <- constant_speed_model(
#'                      screen_distance = 20,
#'                      frame_rate = 60,
#'                      speed = 500,
#'                      attacker_diameter = 50,
#'                      start_distance = 1000)
#'
#' ## Extract the ALT at frame 100
#' get_alt(loom_model,
#'         response_frame = 100)
#'
#' ## Adjust for different viewing distance and apply a response latency
#' get_alt(loom_model,
#'         response_frame = 100,
#'         new_distance = 25,
#'         latency = 0.06)
#'
#' @author Nicholas Carey - \email{nicholascarey@gmail.com}
#'
#' @importFrom utils tail
#'
#' @export

get_alt <-

  function(x,
           response_frame = NULL,
           new_distance = NULL,
           latency = 0){


    # Checks ------------------------------------------------------------------

    ## check class
    if(!any(class(x) %in% c("constant_speed_model", "variable_speed_model", "diameter_model")))
      stop("Input must be an object of class 'constant_speed_model', 'variable_speed_model',
           or 'diameter_model'.")

    if(is.null(response_frame))
      stop("A 'response_frame' is required to extract the ALT and associated data.")

    if(response_frame == 1)
      stop("ALT cannot be extracted from the first frame because it is a derivative and is calculated
           between the response_frame and the previous frame.")

    if(response_frame > tail(x$model$frame, 1))
      stop("The 'response_frame' is greater than the last frame of the animation model.")



    # Diameter model ----------------------------------------------------------

    if(inherits(x, "diameter_model")){

      ## check screen_distance not empty
      if(is.null(new_distance))
        stop("Extracting data from 'diameter_model' objects requires a screen viewing distance.
             This should be entered as 'new_distance'.")

      ## save inputs for inclusion in final output
      original_model <- x
      inputs <- list(
        new_distance = new_distance,
        latency = latency,
        response_frame = response_frame
      )

      ## take out df with frames, animation diameters etc
      adjusted_model <- x$model

      ## set exp parameters
      frame_rate <- x$frame_rate
      screen_dist <- new_distance


      ## latency correction
      response_frame_original <- response_frame
      response_frame_adjusted <- response_frame-(round(frame_rate*latency))

      ## alpha column - visual angle of shape
      adjusted_model$alpha <- 2*(atan((adjusted_model$diam_on_screen/2)/screen_dist))
      adjusted_model$alpha_deg <- rad2deg(adjusted_model$alpha)

      ## da/dt column
      adjusted_model$dadt <- c(NA, diff(adjusted_model$alpha)*frame_rate)
      adjusted_model$dadt_deg <- rad2deg(adjusted_model$dadt)

      ## EXTRACT ALT
      ## from the ADJUSTED FOR LATENCY response frame
      alt <- adjusted_model$dadt[response_frame_adjusted]
      alt_deg <- rad2deg(alt)

      ## organise adjusted model for output
      temp_list <- original_model
      temp_list$model <- adjusted_model
      adjusted_model <- temp_list

      #### OUTPUT

      output <- list(
        alt = alt,
        alt_deg = alt_deg,

        response_frame = response_frame_original,
        response_frame_adjusted = response_frame_adjusted,
        latency_applied = latency,

        distance_perceived = NULL,
        speed_perceived = NULL,
        distance_in_model = NULL,
        speed_in_model = NULL,
        new_distance_applied = inputs$new_distance,

        adjusted_model = adjusted_model,
        original_model = original_model,
        inputs = inputs
      )
    }


    # Constant speed model ----------------------------------------------------

    if(inherits(x, "constant_speed_model")){

      ## save inputs for inclusion in final output
      original_model <- x
      inputs <- list(
        new_distance = new_distance,
        latency = latency,
        response_frame = response_frame
      )


      ## take out df with frames, animation diameters etc
      adjusted_model <- x$model

      ## take out exp parameters
      attacker_diameter <- x$attacker_diameter
      frame_rate <- x$frame_rate

      ## if new distance entered use it
      ifelse(is.null(new_distance),
             screen_dist <- x$screen_distance,
             screen_dist <- new_distance)

      ## latency correction
      response_frame_original <- response_frame
      response_frame_adjusted <- response_frame-(round(frame_rate*latency))

      ## alpha column - visual angle of shape
      adjusted_model$alpha <- 2*(atan((adjusted_model$diam_on_screen/2)/screen_dist))
      adjusted_model$alpha_deg <- rad2deg(adjusted_model$alpha)

      ## da/dt column
      adjusted_model$dadt <- c(NA, diff(adjusted_model$alpha)*frame_rate)
      adjusted_model$dadt_deg <- rad2deg(adjusted_model$dadt)

      ## perceived distance
      adjusted_model$distance_perceived <-
        cos(adjusted_model$alpha/2)*
        (attacker_diameter/2)/(sin((adjusted_model$alpha/2)))

      ## perceived speed
      ## = perceived distance change per s
      adjusted_model$speed_perceived <-
        -1*c(NA, diff(adjusted_model$distance_perceived)*frame_rate)

      ## EXTRACT ALT
      ## from the ADJUSTED FOR LATENCY response frame
      alt <- adjusted_model$dadt[response_frame_adjusted]
      alt_deg <- rad2deg(alt)

      ## get three metrics at the ADJUSTED FOR LATENCY response frame
      alt_perceived <- alt
      distance_perceived <- adjusted_model$distance_perceived[response_frame_adjusted]
      speed_perceived <- adjusted_model$speed_perceived[response_frame_adjusted]

      ## get dist and speed of model at response frame
      distance_in_model <- adjusted_model$distance[response_frame_adjusted]
      speed_in_model <- original_model$speed


      ## organise adjusted model for output
      temp_list <- original_model
      temp_list$model <- adjusted_model
      adjusted_model <- temp_list


      #### OUTPUT

      output <- list(
        alt = alt,
        alt_deg = alt_deg,

        response_frame = response_frame_original,
        latency_applied = latency,
        response_frame_adjusted = response_frame_adjusted,

        distance_perceived = distance_perceived,
        speed_perceived = speed_perceived,
        distance_in_model = distance_in_model,
        speed_in_model = speed_in_model,
        new_distance_applied = inputs$new_distance,

        original_model = original_model,
        adjusted_model = adjusted_model,
        inputs = inputs
      )
    }


    # Variable speed model ----------------------------------------------------

    if(inherits(x, "variable_speed_model")){

      ## save inputs for inclusion in final output
      original_model <- x
      inputs <- list(
        new_distance = new_distance,
        latency = latency,
        response_frame = response_frame
      )


      ## take out df with frames, animation diameters etc
      adjusted_model <- x$model

      ## take out exp parameters
      attacker_diameter <- x$attacker_diameter
      frame_rate <- x$frame_rate

      ## if new distance entered use it
      ifelse(is.null(new_distance),
             screen_dist <- x$screen_distance,
             screen_dist <- new_distance)

      ## latency correction
      response_frame_original <- response_frame
      response_frame_adjusted <- response_frame-(round(frame_rate*latency))

      ## alpha column - visual angle of shape
      adjusted_model$alpha <- 2*(atan((adjusted_model$diam_on_screen/2)/screen_dist))
      adjusted_model$alpha_deg <- rad2deg(adjusted_model$alpha)

      ## da/dt column
      adjusted_model$dadt <- c(NA, diff(adjusted_model$alpha)*frame_rate)
      adjusted_model$dadt_deg <- rad2deg(adjusted_model$dadt)

      ## perceived distance
      adjusted_model$distance_perceived <-
        cos(adjusted_model$alpha/2)*
        (attacker_diameter/2)/(sin((adjusted_model$alpha/2)))

      ## perceived speed
      ## = perceived distance change per s
      adjusted_model$speed_perceived <-
        -1*c(NA, diff(adjusted_model$distance_perceived)*frame_rate)

      ## EXTRACT ALT
      ## from the ADJUSTED FOR LATENCY response frame
      alt <- adjusted_model$dadt[response_frame_adjusted]
      alt_deg <- rad2deg(alt)

      ## get three metrics at the ADJUSTED FOR LATENCY response frame
      alt_perceived <- alt
      distance_perceived <- adjusted_model$distance_perceived[response_frame_adjusted]
      speed_perceived <- adjusted_model$speed_perceived[response_frame_adjusted]

      ## get dist and speed of model at response frame
      distance_in_model <- adjusted_model$distance[response_frame_adjusted]
      ## this is basically only difference to constant speed model - getting
      ## speed at frame rather than the constant speed in model
      speed_in_model <- original_model$speed[response_frame_adjusted]


      ## organise adjusted model for output
      temp_list <- original_model
      temp_list$model <- adjusted_model
      adjusted_model <- temp_list


      #### OUTPUT

      output <- list(
        alt = alt,
        alt_deg = alt_deg,

        response_frame = response_frame_original,
        latency_applied = latency,
        response_frame_adjusted = response_frame_adjusted,

        distance_perceived = distance_perceived,
        speed_perceived = speed_perceived,
        distance_in_model = distance_in_model,
        speed_in_model = speed_in_model,
        new_distance_applied = inputs$new_distance,

        original_model = original_model,
        adjusted_model = adjusted_model,
        inputs = inputs
      )
    }


    ## Assign class
    class(output) <- "get_alt"

    ## Return output
    return(output)

  } #END


